src/core/ddsc/src/dds_publisher.c

```diff
@@ -38,14 +38,30 @@ static dds_return_t dds_publisher_status_validate (uint32_t mask)
   return (mask & ~DDS_PUBLISHER_STATUS_MASK) ? DDS_RETCODE_BAD_PARAMETER : DDS_RETCODE_OK;
 }
 
+static dds_return_t dds_publisher_enable (struct dds_entity *e)
+{
+  struct dds_publisher *pub = (struct dds_publisher *) e;
+
+  assert (dds_entity_kind (e) == DDS_KIND_PUBLISHER);
+
+  /* calling enable on an entity whose factory is not enabled
+   * returns PRECONDITION_NOT_MET */
+  if ((pub->m_entity.m_parent != NULL) && !dds_entity_is_enabled(pub->m_entity.m_parent))
+    return DDS_RETCODE_PRECONDITION_NOT_MET;
+  pub->m_entity.m_flags |= DDS_ENTITY_ENABLED;
```

Comment:
These tests, setting the `ENABLED` flag and propagating to children, are a fundamental part of enabling entities. So in my view, these particular parts of enabling an entity should be part of the generic `dds_enable`, and not be part of the entity kind-specific functions.